development/webpack/utils/git.ts

```diff
@@ -0,0 +1,60 @@
+import { spawnSync } from 'node:child_process';
+import { join, normalize } from 'node:path';
+
+type Commit = {
+  hash: () => string;
+  timestamp: () => number;
+};
+
+/**
+ * Cache variable to store the most recent commit information. This is used to
+ * avoid repeated calls to the git command for performance optimization.
+ */
+const cache: Map<string, Commit> = new Map();
+
+/**
+ * Retrieves the most recent Git commit information, including its short hash
+ * and timestamp. If the information is cached, the cached value is returned to
+ * improve performance. Otherwise, it executes a git command to retrieve the
+ * latest commit's hash and timestamp, caches it, and then returns the
+ * information.
+ *
+ * @param gitDir
+ * @returns The latest commit's hash and timestamp.
+ */
+export function getLatestCommit(
+  gitDir: string = join(__dirname, '../../../.git'),
+): Commit {
+  const cached = cache.get(gitDir);
+  if (cached) return cached;
+
+  // execute the `git` command to get the latest commit's 8 character hash
+  // (`%h` and `--abbrev=8`) and authorship timestamp (seconds since the Unix
+  // epoch)
+  const hashLength = 8;
+  const args = [
+    `--git-dir`,
+    normalize(gitDir),
+    'log',
+    '-1',
+    '--format=%h%at',
+    `--abbrev=${hashLength}`,
+  ] as const;
+  const { stdout } = spawnSync('git', args, {
```

Comment:
I originally implemented this in pure JS in order to not require any external dependencies. It worked and was slightly faster than shelling out to `git`, but when I started writing tests and coming up with some of the more exotic git states the implementation starts getting pretty huge and complicated. There _is_ an open source git implementation that handles _most_ git repo stats (isomorphic git), but it's very slow.